idseq_dag/util/server.py

```diff
@@ -173,6 +185,42 @@ def wait_for_server_ip(service_name,
         # if we had to wait here, that counts toward the rate limit delay
         result = wait_for_server_ip_work(service_name, key_path,
                                          remote_username, environment,
-                                         max_concurrent, chunk_id)
+                                         max_concurrent, chunk_id,
+                                         max_interval_between_describe_instances,
+                                         draining_tag)
         return result
 
+
+def build_job_tag(job_tag_prefix, chunk_id):
+    batch_job_id = os.environ.get('AWS_BATCH_JOB_ID', 'local')
+    job_tag_key = f"{job_tag_prefix}{batch_job_id}_chunk{chunk_id}"
+    job_tag_value = int(time.time())
+    return job_tag_key, job_tag_value
+
+
+def create_tag(instance_iD, tag_key, tag_value):
+    command.execute(f"aws ec2 create-tags --resources {instance_iD} --tags Key={tag_key},Value={tag_value}")
+
+def delete_tag(instance_iD, tag_key):
+    command.execute(f"aws ec2 delete-tags --resources {instance_iD} --tags Key={tag_key}")
+
+
+@contextmanager
+def ASGInstance(service, key_path, remote_username, environment, max_concurrent, chunk_id,
+                max_interval_between_describe_instances=900,
+                job_tag_prefix="RunningIDseqBatchJob_",
+                job_tag_refresh_seconds=600,
+                draining_tag="draining"):
+    instance_ip, instance_iD = wait_for_server_ip(service, key_path, remote_username, environment, max_concurrent, chunk_id,
+        max_interval_between_describe_instances, draining_tag)
+    log.write(f"starting alignment for chunk {chunk_id} on {service} server {instance_ip}")
+    job_tag_key, job_tag_value = build_job_tag(job_tag_prefix, chunk_id)
+    t = PeriodicThread(target=create_tag, sleep_seconds=job_tag_refresh_seconds,
+                       args=(instance_iD, job_tag_key, job_tag_value))
+    t.start()
+
+    yield instance_ip
+
+    t.stop()
```

Comment:
My reading of the documentation is that you need a try...finally block here so that you will correctly delete the tag even if the SSH command crashes.

```
try:
    yield instance_ip
finally:
    all the rest of the code
```

Here is an excerpt from the documentation example
```
from contextlib import contextmanager

@contextmanager
def managed_resource(*args, **kwds):
    # Code to acquire resource, e.g.:
    resource = acquire_resource(*args, **kwds)
    try:
        yield resource
    finally:
        # Code to release resource, e.g.:
        release_resource(resource)

>>> with managed_resource(timeout=3600) as resource:
...     # Resource is released at the end of this block,
...     # even if code in the block raises an exception
```

Reply:
Thanks